### PR TITLE
[Prototype form component] Form II -- jsonable!

### PR DESCRIPTION
**Pull request by kenning**
Same as https://github.com/getethos/ethos-design-system/pull/22 except now everything except the onSubmit of the form is defined with 100% json-compatible syntax. It's pretty readable too

### CODE REVIEW

File: src/components/Form/Form.js [hunk 1]
```diff
@@ -23,20 +25,6 @@ export function Form({ children, config, render }) {
   ] = useFormState(initialValues)
 
   // Validates form, sets button to disabled if there's an error
-  function inputNamePropsTransformer(inputName) {
```
Comment (kenning) [inline src/components/Form/Form.js]: This basically turns into "input" function below

File: src/components/Form/ValidatorMap.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import validateMinMaxFactory from '../../validators/validateMinMax'
+import validateTruthy from '../../validators/validateTruthy'
+
+export default function GenerateValidator(validatorName, args) {
+  switch (validatorName) {
+    case 'truthy':
+      return validateTruthy
+    case 'minMax':
+      return validateMinMaxFactory.apply(null, args)
+    case 'exampleEvenNumber':
```
Comment (kenning) [inline src/components/Form/ValidatorMap.js]: Throwaway code for the purposes of illustrating what this setup can do.

File: src/components/Form/ComponentMap.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+import { TextInput } from '../index'
+
+export default function GenerateComponent(componentName, props) {
```
Comment (roblevintennis) [inline src/components/Form/ComponentMap.js]: Ok, I'd call this a `ComponentGenerator` fwiw, but, I think for this sort of factory thing to be useful you MUST keep a library of widgets that matches your components. So it violates open/closed principle (google it) in that you'll have to keep it up to date by adding/removing components from your switch state here (I presume this is incomplete for POC purposes. But, it _would_ need to potentially support all components.

Comment (roblevintennis) [reply src/components/Form/ComponentMap.js]: Think about how shitty kitchensink started being to keep up to date for example-- that's the sort of high maintenance I see this becoming

File: src/components/Form/ComponentMap.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  switch (componentName) {
+    case 'TextInput':
+      return <TextInput {...props} />
+    default:
+      throw new Error('whoopsie daisy')
```
Comment (roblevintennis) [inline src/components/Form/ComponentMap.js]: So this confirms to my mind that your intent is you can only ask for components that exist. So my assertion above about having to maintain the list seems accurate.

File: src/components/Form/Form.js [hunk 5]
```diff
@@ -58,21 +46,33 @@ export function Form({ children, config, render }) {
...
+          "\n\nBut try again, it's random"
       )
     } catch (e) {
       setFormErrorMessage('Something bad happened: ' + e.toString())
     }
   }
 
+  function input(inputName) {
+    const inputConfig = config.inputs[inputName]
+    return ComponentMap(inputConfig.componentName, {
```
Comment (roblevintennis) [inline src/components/Form/Form.js]: Ok, I see how your using this. So...

`ComponentMap` to my mind should be controlled by the client. This way, the client can assemble just the "UI widgets" it needs and circumnavigates my concerns about maintaining a big switch case list of components. So, I have a form and I want a `TextInput` and `ButtonSelectGroup`...in that case, I can assemble my own component map on `config.map` (and even validators with `config.validators`).

I think you could do ^ without really changing the spirit of what you're doing here, no?

If I had any mandate, I'd say inverting the control on the maps and validators is my biggest concern.

File: src/components/Form/Form.js [hunk 6]
```diff
@@ -58,21 +46,33 @@ export function Form({ children, config, render }) {
...
+          JSON.stringify(getInputValues()) +
+          "\n\nBut try again, it's random"
       )
     } catch (e) {
       setFormErrorMessage('Something bad happened: ' + e.toString())
     }
   }
 
+  function input(inputName) {
+    const inputConfig = config.inputs[inputName]
```
Comment (roblevintennis) [inline src/components/Form/Form.js]: might as well start using the vernacular fields for all of this no? Or is that v2b? :)

Comment (roblevintennis) [reply src/components/Form/Form.js]: But that's the direction you have in mind for later yah??

File: src/components/Form/ValidatorMap.js (hunk 2 above)
Comment (roblevintennis) [reply src/components/Form/ValidatorMap.js]: Yeah, I'd say my comments for the "mappers" and "validators" can be considered the same. I don't want us to have to manage/maintain all these. I think it's enough for us to provide some boiler-plate validators, but this generation/builder type of stuff belongs higher up layer and on the client side to my mind.

Comment (kenning) [reply src/components/Form/ValidatorMap.js]: sounds good

File: src/components/Form/ComponentMap.js (hunk 3 above)
Comment (kenning) [reply src/components/Form/ComponentMap.js]: yep will push changes

**Pull request closed without merging**